File: src/api/client/repository/interface.ts
import { FileDescAPIResponse } from 'src/api/client/model';
import { TGFSFileRef } from 'src/model/directory';
import { TGFSFile } from 'src/model/file';
import { TGFSMetadata } from 'src/model/metadata';

export interface IFDRepository {
  save(fd: TGFSFile, messageId?: number): Promise<FileDescAPIResponse>;
  get(fr: TGFSFileRef): Promise<TGFSFile>;
}

export interface IMetaDataRepository {
  save(metaData: TGFSMetadata): Promise<number>;
  get(): Promise<TGFSMetadata>;
}

File: tgfs/api/client/repository/interface.py
from abc import ABCMeta, abstractmethod
from typing import Optional

from tgfs.api.client.api.model import FileDescAPIResponse
from tgfs.model.file import TGFSFile
from tgfs.model.directory import TGFSFileRef
from tgfs.model.metadata import TGFSMetadata


class IFDRepository(metaclass=ABCMeta):
    @abstractmethod
    async def save(
        self, fd: TGFSFile, message_id: Optional[int] = None
    ) -> FileDescAPIResponse:
        pass

    @abstractmethod
    async def get(self, fr: TGFSFileRef) -> TGFSFile:
        pass


class IMetaDataRepository(metaclass=ABCMeta):
    @abstractmethod
    async def save(self, metadata: TGFSMetadata) -> int:
        pass

    @abstractmethod
    async def get(self) -> TGFSMetadata:
        pass
